File: java/com/lc/lc6433.java
package lc;

import java.util.Deque;
import java.util.LinkedList;

/**
 * @program: exercise
 * @description: lc6433
 * @author: 郭晨旭
 * @create: 2023-05-14 10:53
 * @version: 1.0
 **/
public class lc6433 {
    public int maxMoves(int[][] grid) {
        int moves = 0;
        int m = grid.length, n = grid[0].length;
        int[][] vis = new int[m][n];
        Deque<int[]> queue = new LinkedList<>();
        for (int i = 0; i < m; ++i) {
            queue.add(new int[] { i, 0, 0 });
        }
        int[][] step = new int[][] { { 1, 1 }, { 0, 1 }, { -1, 1 } };
        while (!queue.isEmpty()) {
            int[] now = queue.poll();
            for (int i = 0; i < 3; ++i) {
                int nx = now[0] + step[i][0];
                int ny = now[1] + step[i][1];
                if (ok(m, n, nx, ny) && grid[nx][ny] > grid[now[0]][now[1]]) {
                    if (now[2] + 1 > vis[nx][ny]) {
                        queue.offer(new int[] { nx, ny, now[2] + 1 });
                        vis[nx][ny] = now[2] + 1;
                    }
                } else {
                    moves = Math.max(moves, now[2]);
                }
            }
        }
        return moves;
    }

    public boolean ok(int m, int n, int x, int y) {
        return x >= 0 && x < m && y >= 0 && y < n;
    }
}

File: go/lc6433.go
package main

import (
	"container/list"
	"math"
)

func ok(m, n, x, y int) bool {
	return x >= 0 && x < m && y >= 0 && y < n
}

func maxMoves(grid [][]int) int {
	moves := 0
	m := len(grid)
	n := len(grid[0])
	vis := make([][]int, m)
	for i := range vis {
		vis[i] = make([]int, n)
	}
	queue := list.New()
	for i := 0; i < m; i++ {
		queue.PushBack([3]int{i, 0, 0})
		vis[i][0] = 0
	}
	step := [3][2]int{{1, 1}, {0, 1}, {-1, 1}}
	for queue.Len() > 0 {
		now := queue.Front().Value.([3]int)
		queue.Remove(queue.Front())
		for _, s := range step {
			nx := now[0] + s[0]
			ny := now[1] + s[1]
			if ok(m, n, nx, ny) && grid[nx][ny] > grid[now[0]][now[1]] {
				if now[2]+1 > vis[nx][ny] {
					queue.PushBack([3]int{nx, ny, now[2] + 1})
					vis[nx][ny] = now[2] + 1
				}
			} else {
				moves = int(math.Max(float64(moves), float64(now[2])))
			}
		}
	}
	return moves
}
